tests/optim/test_layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,220 @@
+                    if hasattr(param, "grad"):
+                        logging.debug("testing equality for %s.%s" % (module_name, param_name))
+                        result.append(param.grad)
+        return result
+
+    for elt in zip(get_params_with_grad(vanilla_model), get_params_with_grad(scaled_model)):
+        assert torch.allclose(elt[0], elt[1])
+
+
+# Test: convolutional network
+class SimpleConvNet(nn.Module):
+    def __init__(self):
+        torch.manual_seed(24)
+        super(SimpleConvNet, self).__init__()
+        self.conv1 = nn.Conv2d(3, 6, 5)
+        self.relu1 = nn.ReLU()
+        self.pool1 = nn.MaxPool2d(2, 2)
+        self.conv2 = nn.Conv2d(6, 16, 5)
+        self.relu2 = nn.ReLU()
+        self.pool2 = nn.MaxPool2d(2, 2)
+        self.fc1 = nn.Linear(16 * 5 * 5, 120)
+        self.relu3 = nn.ReLU()
+        self.fc2 = nn.Linear(120, 84)
+        self.relu4 = nn.ReLU()
+        self.fc3 = nn.Linear(84, 10)
+        self.identity = nn.Identity()
+
+    def forward(self, x):
+        out = self.conv1(x)
+        out = self.relu1(out)
+        out = self.pool1(out)
+        out = self.conv2(out)
+        out = self.relu2(out)
+        out = self.pool2(out)
+        out = torch.flatten(out, 1)  # flatten all dimensions except batch
+        out = self.fc1(out)
+        out = self.relu3(out)
+        out = self.fc2(out)
+        out = self.relu4(out)
+        out = self.fc3(out)
+        out = self.identity(out)
+        return out
+
+
+def train_vision_model(model: SimpleConvNet, per_layer_scaling=False):
+    loss_fn = nn.CrossEntropyLoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
+
+    if torch.cuda.is_available():
+        model.cuda()
+
+    train_ds_loader = load_data("vision_model")
+    model.train()
+
+    layer_scale_dict = {"conv1": 128, "conv2": 256, "fc1": 512, "fc2": 1024, "fc3": 8192} if per_layer_scaling else {}
+    layer_scaler = LayerwiseGradientScaler(model, layer_scale_dict)
+
+    for _ in range(2):
+        for img, lbl in train_ds_loader:
+            if torch.cuda.is_available():
+                img = img.cuda()
+                lbl = lbl.cuda()
+
+            optimizer.zero_grad()
+            layer_scaler.scale()
+
+            predict = model(img)
+            loss = loss_fn(predict, lbl)
+
+            loss.backward()
+
+            layer_scaler.unscale()
+            layer_scaler.step(optimizer)
+    return model
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="cuda required")
+def test_vision_model() -> None:
+    # Remove randomness from various sources while testing.
+    torch.use_deterministic_algorithms(True)  # type: ignore
+    # set environment variable in CircleCI for test to pass: CUBLAS_WORKSPACE_CONFIG = :4096:8
+
+    m1 = SimpleConvNet()
+    m2 = SimpleConvNet()
+
+    vision_model = train_vision_model(m1, False)
+    scaled_vision_model = train_vision_model(m2, True)
```

Comment:
I am guessing that without the layerwise scaling scaled_vision_model would have inf grads? is it worth adding a test for that?

Reply:
At present neither of the models encounter any infs/nans. It is tricky to build a test case which uses mixed precision and has inf in the gradient which stabilize with layerwise gradient scaling. These tests check that the correct scaling factor is being applied on each layer and that unscaling is working correctly.

Reply:
The test_linear_model does use mixed precision as we wrap the forward pass with the `autocast` context.

Reply:
Discussed offline: there is a test where we see infs that can be used here. 

I agree that is tricky to build a tests case but unless we do we cannot test if this works. I would also suggest that we use larger models if needed to simulate this behavior as a one off test case. 

Let us add this as a TODO. Shall we move this to experimental/ in the meantime? WDYT?

Reply:
I'll document this and moving it to experimental.